src/mds/Locker.cc

```diff
@@ -2223,6 +2223,14 @@ Capability* Locker::issue_new_caps(CInode *in,
       cap->set_wanted(cap->wanted() | my_want);
     }
   }
+
+  if (!mdr->session->info.has_feature(CEPHFS_FEATURE_MAX_XATTR_PAIRS_SIZE)) {
+    dout(10) << "No max xattr support on the client; dropping Xx caps: "
+             << ccap_string(cap->wanted()) << " => "
+             << ccap_string(cap->wanted() & ~CEPH_CAP_XATTR_EXCL) << dendl;
+    cap->set_wanted(cap->wanted() & ~CEPH_CAP_XATTR_EXCL);
+  }
```

Comment:
Please fix the `Xx` issue in one separate commit.

I am afraid this won't work. The MDS will always try to issue more than wanted if possible, please see https://github.com/ceph/ceph/blob/main/src/mds/CInode.cc#L3923.

IMO you should do this in `Locker::issue_caps()` and `CInode::get_caps_allowed_for_client()`. Possibly you could add one helper in `Session`, such as `Session::get_caps_disallowed()`.

Reply:
Hmm.. ok, I'll have a closer look.  And adding an helper to `Session` makes sense indeed if we're doing this in different places.